daemon/graphdriver/btrfs/btrfs.go

```diff
@@ -179,148 +203,355 @@ func free(p *C.char) {
+			dirid := C._le64toh(ref.dirid)
+
+			// Search relative path to subvol parent dir
+			var inoArgs C.struct_btrfs_ioctl_ino_lookup_args
+			inoArgs.treeid = subvolID
+			inoArgs.objectid = dirid
+			_, _, errno := unix.Syscall(unix.SYS_IOCTL, getDirFd(dir), C.BTRFS_IOC_INO_LOOKUP,
+				uintptr(unsafe.Pointer(&inoArgs)))
+			if errno == syscall.ENAMETOOLONG {
+				// Path too long so save it for future walk
+				distantDescendants[dirid] = append(distantDescendants[dirid], subvolData{childName, sh.objectid})
+			} else if errno != 0 {
+				return fmt.Errorf("Failed to search relative path from %s to %s: %v", name, childName, errno.Error())
+			} else {
+				relPath := C.GoString((*C.char)(unsafe.Pointer(&inoArgs.name)))
+				if err := subvolDelete(path.Join(fullPath, relPath), childName, sh.objectid, quotaEnabled); err != nil {
+					return err
+				}
 			}
-		} else {
-			logrus.WithField("storage-driver", "btrfs").Errorf("Failed to lookup btrfs qgroup for %s: %v", fullPath, err.Error())
+		}
+		// Increase search key by one, to read the next item, if we can.
+		if btrfsIoctlSearchArgsInc(&args) {
+			break
 		}
 	}
 
-	// all subvolumes have been removed
-	// now remove the one originally passed in
-	for i, c := range []byte(name) {
-		args.name[i] = C.char(c)
+	// If there's any subvolumes whose path is too long, process it now
+	if len(distantDescendants) > 0 {
+		var walk func(*C.DIR, string) error
+		walk = func(pdir *C.DIR, ppath string) error {
+			ent, errno := C.readdir(pdir)
+			for ; ent != nil; ent, errno = C.readdir(pdir) {
+				if ent.d_type == C.DT_DIR {
+					dname := C.GoString((*C.char)(unsafe.Pointer(&ent.d_name)))
+					// Check if dir contain subvols
+					if childs, ok := distantDescendants[C.__u64(ent.d_ino)]; ok {
+						dpath := path.Join(ppath, dname)
+						for _, child := range childs {
+							if err := subvolDelete(dpath, child.name, child.id, quotaEnabled); err != nil {
+								return fmt.Errorf("Failed to delete btrfs subvolume %+v: %v", child, err)
+							}
+						}
+						delete(distantDescendants, C.__u64(ent.d_ino))
+					}
+					// Go deeper
+					if dname != "." && dname != ".." {
+						var oflags C.int = (C.O_RDONLY | C.O_NDELAY | C.O_DIRECTORY | C.O_CLOEXEC | C.O_NOFOLLOW)
+						dirfd, err := C._openat(C.dirfd(pdir), (*C.char)(unsafe.Pointer(&ent.d_name)), oflags)
+						if dirfd < 0 {
+							logrus.Errorf("Can't open dir %s: %v", dname, err)
+							continue
+						}
+						cdir, err := C.fdopendir(dirfd)
+						if cdir == nil {
+							C.close(dirfd)
+							logrus.Errorf("Can't get DIR from fd %d for %s: %v", dirfd, dname, err)
+							continue
+						}
+						defer closeDir(cdir)
+
+						if err := walk(cdir, path.Join(ppath, dname)); err != nil {
+							return err
+						}
+					}
+				}
+			}
+			if errno != nil {
+				logrus.Errorf("Error while walk at %s: %v", ppath, errno)
+			}
+			return nil
+		}
```

Comment:
I see no reason Cgo would be needed to implement `walk`. A pure-Go implementation would be more readable, and potentially faster without the Cgo thunk overhead. I have illustrated this with a 1:1 conversion, though there is lots of room for improvement, such as leveraging `filepath.WalkDir` or eliding the integer typecasts by keying the `distantDescendants` map by `uint64`.
```suggestion
		walk = func(pdir *os.File, ppath string) error {
			ents, err := pdir.ReadDir(1)
			for ; err == nil; ents, err = pdir.ReadDir(1) {
				ent := ents[0]
				if ent.IsDir() {
					// Check if dir contain subvols
					info, err := ent.Info()
					if err != nil {
						if errors.Is(err, os.ErrNotExist) {
							continue
						}
						return err
					}
					ino := info.Sys().(*syscall.Stat_t).Ino
					if childs, ok := distantDescendants[C.__u64(ino)]; ok {
						dpath := path.Join(ppath, ent.Name())
						for _, child := range childs {
							if err := subvolDelete(dpath, child.name, child.id, quotaEnabled); err != nil {
								return fmt.Errorf("Failed to delete btrfs subvolume %+v: %v", child, err)
							}
						}
						delete(distantDescendants, C.__u64(ino))
					}
					// Go deeper
					const oflags = (unix.O_RDONLY | unix..O_NDELAY | unix.O_DIRECTORY | unix.O_CLOEXEC | unix.O_NOFOLLOW)
					dirfd, err := unix.Openat(pdir.Fd(), ent.Name(), oflags, 0)
					if err != nil {
						logrus.Errorf("Can't open dir %s: %v", ent.Name(), err)
						continue
					}
					cdir := os.NewFile(dirfd, ent.Name())
					defer cdir.Close()

					if err := walk(cdir, path.Join(ppath, ent.Name())); err != nil {
						return err
					}
				}
			}
			if err == io.EOF {
				return nil
			}
			return err
		}
```